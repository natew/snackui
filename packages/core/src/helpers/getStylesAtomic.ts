import {
  StyleObject,
  invertMapTransformKeys,
  mergeTransform,
  stylePropsTransform,
} from '@tamagui/helpers'
import { ViewStyle } from 'react-native'

import { getConfig } from '../conf'
import { rnw } from '../constants/rnw'
import { isVariable } from '../createVariable'

// NOTE: going to refactor and merge getSplitStyles + getAtomicStyles into one
// before i can do that need to not rely on rnw for generateAtomicStyles
// will change that to not be external at all and just logically do it in the
// merged getSplitStyles/getAtomicStyles file in one loop.

export type ViewStyleWithPseudos = ViewStyle & {
  hoverStyle?: ViewStyle
  pressStyle?: ViewStyle
  focusStyle?: ViewStyle
}

type AtomicStyleOptions = {
  splitTransforms?: boolean
}

export function getStylesAtomic(stylesIn: ViewStyleWithPseudos, options: AtomicStyleOptions = {}) {
  const { hoverStyle, pressStyle, focusStyle, ...base } = stylesIn
  let res: StyleObject[] = []

  // *1 order matched to *0
  for (const [index, style] of [hoverStyle, pressStyle, focusStyle, base].entries()) {
    if (!style || !Object.keys(style).length) {
      continue
    }
    const pseudo = pseudosOrdered[index]
    for (const skey in style) {
      const val = style[skey]
      if (isVariable(val)) {
        style[skey] = val.toString()
      }
      if (stylePropsTransform[skey]) {
        delete style[skey]
        mergeTransform(style, skey, val)
      }
    }
    res = [...res, ...getAtomicStyle(style, pseudo, options)]
  }

  return res
}

const generateAtomicStyles = (style: ViewStyle) => {
  return rnw.atomic(rnw.createCompileableStyle(rnw.createReactDOMStyle(style))) as {
    [key: string]: StyleObject
  }
}

let reversedShorthands: Record<string, string> | null = null

function getAtomicStyle(
  style: ViewStyle,
  pseudo: { name: string; priority: number } | undefined,
  options: AtomicStyleOptions
): StyleObject[] {
  if (style == null || typeof style !== 'object') {
    throw new Error(`Wrong style type: "${typeof style}": ${style}`)
  }

  // why is this diff from react-native-web!? need to figure out
  for (const key in borderDefaults) {
    if (key in style) {
      style[borderDefaults[key]] = style[borderDefaults[key]] ?? 'solid'
    }
  }

  let atomicStyles: { [key: string]: StyleObject & { transformProperty?: string } } = {}

  if (options.splitTransforms && style.transform) {
    let { transform, ...rest } = style
    Object.assign(atomicStyles, generateAtomicStyles(rest))
    for (const t of transform) {
      const tKey = Object.keys(t)[0]
      const transformProperty = invertMapTransformKeys[tKey] || tKey
      const out = generateAtomicStyles({ transform: [t] })
      const key = Object.keys(out)[0]
      atomicStyles[key] = {
        ...out[key],
        transformProperty,
      }
    }
  } else {
    // TODO we can do this all in one loop likely inside getSplitStyles in existing loop and avoid O(n^3)...
    atomicStyles = generateAtomicStyles(style)
  }

  if (!reversedShorthands) {
    reversedShorthands = {}
    const conf = getConfig()
    if (conf.shorthands) {
      for (const key in conf.shorthands) {
        reversedShorthands[conf.shorthands[key]] = key
      }
    }
  }

  // TODO ... and then also avoid this loop! n^4
  return Object.keys(atomicStyles).map((key) => {
    const val = atomicStyles[key]
    // r-transform-1ns13n
    const [_, a, b] = val.identifier.split('-')
    // dev mode its "r-transform-1ns13n", prod its "r-1ns13n", normalize them
    const hash = b || a
    // pseudos have a `--` to be easier to find with concatClassNames
    const psuedoPrefix = pseudo ? `0${pseudo.name}-` : ''
    const shortProp = reversedShorthands![val.property] || val.property
    const identifier = `_${shortProp}-${psuedoPrefix}${hash}`
    const className = `.${identifier}`
    const rules = val.rules.map((rule) => {
      if (pseudo) {
        const psuedoPrefixSelect = [...Array(pseudo.priority)].map(() => ':root').join('') + ' '
        let res = rule
          .replace(`.${val.identifier}`, `${psuedoPrefixSelect} ${className}`)
          .replace('{', `:${pseudo.name}{`)

        if (pseudo.name === 'hover') {
          // hover styles need to be conditional
          // perhaps this can be generalized but for now lets just shortcut
          // and hardcode for hover styles, if we need to later we can
          // WEIRD SYNTAX, SEE:
          //   https://stackoverflow.com/questions/40532204/media-query-for-devices-supporting-hover
          res = `@media not all and (hover: none) { ${res} }`
        }
        return res
      }
      return rule.replace(`.${val.identifier}`, className).replace(importantRegex, '')
    })

    const result: StyleObject = {
      property: val.transformProperty || val.property,
      value: val.value,
      identifier,
      className,
      rules,
    }

    return result
  })
}

const importantRegex = /\!important*/g

const borderDefaults = {
  borderWidth: 'borderStyle',
  borderBottomWidth: 'borderBottomStyle',
  borderTopWidth: 'borderTopStyle',
  borderLeftWidth: 'borderLeftStyle',
  borderRightWidth: 'borderRightStyle',
}

// *0 order matches to *1
export const pseudos = {
  hoverStyle: {
    name: 'hover',
    priority: 1,
  },
  pressStyle: {
    name: 'active',
    priority: 2,
  },
  focusStyle: {
    name: 'focus',
    priority: 3,
  },
}

const pseudosOrdered = Object.values(pseudos)
