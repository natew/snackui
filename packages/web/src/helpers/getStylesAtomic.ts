/**
 * Some parts adapted from react-native-web
 * Copyright (c) Nicolas Gallagher licensed under the MIT license.
 */

import { StyleObject, simpleHash } from '@tamagui/helpers'
import type { TextStyle, ViewStyle } from 'react-native'

import { getConfig } from '../config'
import { TamaguiInternalConfig } from '../types'
import { defaultOffset } from './expandStyles'
import { normalizeValueWithProperty } from './normalizeValueWithProperty.js'
import { PseudoDescriptor, pseudoDescriptors } from './pseudoDescriptors'

// refactor this file away next...

type ViewOrTextStyle = ViewStyle | TextStyle

export type ViewStyleWithPseudos = ViewOrTextStyle & {
  hoverStyle?: ViewOrTextStyle
  pressStyle?: ViewOrTextStyle
  focusStyle?: ViewOrTextStyle
}

// matching order of the below *0
const pseudosOrdered = [
  pseudoDescriptors.hoverStyle,
  pseudoDescriptors.pressStyle,
  pseudoDescriptors.focusStyle,
]

export function getStylesAtomic(stylesIn: ViewStyleWithPseudos) {
  // performance optimization
  if (!(stylesIn.hoverStyle || stylesIn.pressStyle || stylesIn.focusStyle)) {
    return getAtomicStyle(stylesIn)
  }

  // only for pseudos
  const { hoverStyle, pressStyle, focusStyle, ...base } = stylesIn
  let res: StyleObject[] = []
  // *1 order matched to *0
  for (const [index, style] of [hoverStyle, pressStyle, focusStyle, base].entries()) {
    if (!style) continue
    const pseudo = pseudosOrdered[index]
    res = [...res, ...getAtomicStyle(style, pseudo)]
  }
  return res
}

const cache = new Map()

export function getAtomicStyle(
  style: ViewOrTextStyle,
  pseudo?: PseudoDescriptor
): StyleObject[] {
  if (!style) return []
  const key = JSON.stringify(style) + (pseudo ? JSON.stringify(pseudo) : '')
  if (cache.has(key)) {
    return cache.get(key)
  }
  if (cache.size > 800) {
    cache.clear()
  }
  if (process.env.NODE_ENV === 'development') {
    if (!style || typeof style !== 'object') {
      throw new Error(`Wrong style type: "${typeof style}": ${style}`)
    }
  }
  const out = generateAtomicStyles(style, pseudo)
  cache.set(key, out)
  return out
}

let conf: TamaguiInternalConfig

const generateAtomicStyles = (
  styleIn: ViewStyle & TextStyle,
  pseudo?: PseudoDescriptor
): StyleObject[] => {
  conf = conf || getConfig()

  // were converting to css styles
  const style = styleIn as Record<string, string | null | undefined>

  // transform
  if (style.transform && Array.isArray(style.transform)) {
    style.transform = style.transform
      .map(
        // { scale: 2 } => 'scale(2)'
        // { translateX: 20 } => 'translateX(20px)'
        // { matrix: [1,2,3,4,5,6] } => 'matrix(1,2,3,4,5,6)'
        (transform) => {
          const type = Object.keys(transform)[0]
          const value = transform[type]
          if (type === 'matrix' || type === 'matrix3d') {
            return `${type}(${value.join(',')})`
          }
          return `${type}(${normalizeValueWithProperty(value, type)})`
        }
      )
      .join(' ')
  }

  styleToCSS(style)

  const out: StyleObject[] = []
  for (const key in style) {
    const value = style[key]
    if (value == null || value == undefined) continue
    const hash = presetHashes[value]
      ? value
      : typeof value === 'string'
      ? simpleHash(value)
      : `${value}`.replace('.', 'dot')

    const pseudoPrefix = pseudo ? `0${pseudo.name}-` : ''
    const shortProp = conf.inverseShorthands[key] || key
    const identifier = `_${shortProp}-${pseudoPrefix}${hash}`
    const rules = createAtomicRules(identifier, key, value, pseudo)
    const styleObject: StyleObject = {
      property: key,
      pseudo: pseudo?.name as any,
      value,
      identifier,
      rules,
    }
    out.push(styleObject)
  }

  return out
}

const presetHashes = {
  none: true,
}

const n = normalizeValueWithProperty

export function styleToCSS(style: Record<string, any>) {
  // box-shadow
  const { shadowOffset, shadowRadius, shadowColor } = style
  if (style.shadowRadius !== undefined) {
    const offset = shadowOffset || defaultOffset
    const shadow = `${n(offset.width)} ${n(offset.height)} ${n(
      shadowRadius
    )} ${shadowColor}`
    style.boxShadow = style.boxShadow ? `${style.boxShadow}, ${shadow}` : shadow
    style.shadowOffset = undefined
    style.shadowRadius = undefined
    style.shadowColor = undefined
  }

  // text-shadow
  const { textShadowColor, textShadowOffset, textShadowRadius } = style
  if (textShadowColor || textShadowOffset || textShadowRadius) {
    const { height, width } = textShadowOffset || defaultOffset
    const radius = textShadowRadius || 0
    const color = normalizeValueWithProperty(textShadowColor, 'textShadowColor')
    if (color && (height !== 0 || width !== 0 || radius !== 0)) {
      const blurRadius = normalizeValueWithProperty(radius)
      const offsetX = normalizeValueWithProperty(width)
      const offsetY = normalizeValueWithProperty(height)
      style.textShadow = `${offsetX} ${offsetY} ${blurRadius} ${color}`
    }
    style.textShadowColor = undefined
    style.textShadowOffset = undefined
    style.textShadowRadius = undefined
  }
}

function createDeclarationBlock(style: Record<string, any>, important = false) {
  let next = ''
  for (const key in style) {
    const prop = hyphenateStyleName(key)
    const value = style[key]
    next += `${prop}:${value}${important ? ' !important' : ''};`
  }
  return `{${next}}`
}

const hcache = {}
const toHyphenLower = (match: string) => `-${match.toLowerCase()}`
const hyphenateStyleName = (key: string) => {
  let val = hcache[key]
  if (val) return val
  hcache[key] = val = key.replace(/[A-Z]/g, toHyphenLower)
  return val
}

const pseudoSelectorPrefixes = (() => {
  const res: Record<string, string> = {}
  for (const key in pseudoDescriptors) {
    const pseudo = pseudoDescriptors[key]
    res[pseudo.name] = `${[...Array(pseudo.priority)].map(() => ':root').join('')} `
  }
  return res
})()

function createAtomicRules(
  identifier: string,
  property: string,
  value: any,
  pseudo?: PseudoDescriptor
): string[] {
  const selector = pseudo
    ? `${pseudoSelectorPrefixes[pseudo.name]} .${identifier}:${pseudo.name}`
    : `.${identifier}`
  const important = !!pseudo

  let rules: string[] = []

  // Handle non-standard properties and object values that require multiple
  // CSS rules to be created.
  switch (property) {
    // Equivalent to using '::placeholder'
    case 'placeholderTextColor': {
      const block = createDeclarationBlock({ color: value, opacity: 1 }, important)
      rules.push(`${selector}::placeholder${block}`)
      break
    }

    // all webkit prefixed rules
    case 'backgroundClip':
    case 'userSelect': {
      const propertyCapitalized = `${property[0].toUpperCase()}${property.slice(1)}`
      const webkitProperty = `Webkit${propertyCapitalized}`
      const block = createDeclarationBlock(
        { [property]: value, [webkitProperty]: value },
        important
      )
      rules.push(`${selector}${block}`)
      break
    }

    // Polyfill for additional 'pointer-events' values
    case 'pointerEvents': {
      let finalValue = value
      if (value === 'auto' || value === 'box-only') {
        finalValue = 'auto'
        if (value === 'box-only') {
          rules.push(`${selector}>*${boxOnly}`)
        }
      } else if (value === 'none' || value === 'box-none') {
        finalValue = 'none'
        if (value === 'box-none') {
          rules.push(`${selector}>*${boxNone}`)
        }
      }
      const block = createDeclarationBlock({ pointerEvents: finalValue }, true)
      rules.push(`${selector}${block}`)
      break
    }

    default: {
      const block = createDeclarationBlock({ [property]: value }, important)
      rules.push(`${selector}${block}`)
      break
    }
  }

  // hover styles need to be conditional
  // perhaps this can be generalized but for now lets just shortcut
  // and hardcode for hover styles, if we need to later we can
  // WEIRD SYNTAX, SEE:
  //   https://stackoverflow.com/questions/40532204/media-query-for-devices-supporting-hover
  if (pseudo && pseudo.name === 'hover') {
    rules = rules.map((r) => `@media not all and (hover: none) { ${r} }`)
  }

  return rules
}

const boxNone = createDeclarationBlock({ pointerEvents: 'auto' }, true)
const boxOnly = createDeclarationBlock({ pointerEvents: 'none' }, true)
