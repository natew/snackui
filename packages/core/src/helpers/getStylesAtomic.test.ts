import { expect, test } from 'vitest'

import { getStylesAtomic } from './getStylesAtomic'

test(`should expand webkit user-select`, () => {
  expect(
    getStylesAtomic({
      // @ts-ignore
      userSelect: 'none',
    })
  ).toMatchSnapshot()
})
