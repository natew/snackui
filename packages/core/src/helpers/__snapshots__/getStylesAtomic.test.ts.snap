// Vitest Snapshot v1

exports[`should expand webkit user-select 1`] = `
[
  {
    "identifier": "_userSelect-none",
    "property": "userSelect",
    "pseudo": undefined,
    "rules": [
      "._userSelect-none{user-select:none;-webkit-user-select:none;}",
    ],
    "value": "none",
  },
]
`;
